### PR TITLE
[Remote Store] Update Translog Metadata file name

### PR DESCRIPTION
**Pull request by gbbafna**
### Description

We are changing translog metadata file name to  

<inverted_primary_term>-- <inverted_generation>--<inverted_timestamp>--<version>

This would enable us to use lexicographic sort given by remote stores and find the latest metadata in just one list call, instead of doing pagination .

With this change, we are also deleting the older metadata files with every cleanup operation. 

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/remote/RemoteStoreUtils.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package org.opensearch.index.remote;
+
+import java.util.Arrays;
+
+/**
+ * Utils for remote store
+ */
+public class RemoteStoreUtils {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/remote/RemoteStoreUtils.java:16]: nit: Call it PrefixConverter

File: server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java [hunk 2]
```diff
@@ -575,8 +536,10 @@ public Map<String, UploadedSegmentMetadata> getSegmentsUploadedToRemoteStore(lon
      * @throws IOException in case of I/O error while reading from / writing to remote segment store
      */
     public void deleteStaleSegments(int lastNMetadataFilesToKeep) throws IOException {
-        Collection<String> metadataFiles = remoteMetadataDirectory.listFilesByPrefix(MetadataFilenameUtils.METADATA_PREFIX);
-        List<String> sortedMetadataFileList = metadataFiles.stream().sorted(METADATA_FILENAME_COMPARATOR).collect(Collectors.toList());
+        List<String> sortedMetadataFileList = remoteMetadataDirectory.listFilesByPrefixInLexicographicOrder(
+            MetadataFilenameUtils.METADATA_PREFIX,
+            Integer.MAX_VALUE
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java:541]: Long?

File: server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java [hunk 3]
```diff
@@ -587,8 +550,8 @@ public void deleteStaleSegments(int lastNMetadataFilesToKeep) throws IOException
         }
 
         List<String> metadataFilesEligibleToDelete = sortedMetadataFileList.subList(
-            0,
-            sortedMetadataFileList.size() - lastNMetadataFilesToKeep
+            lastNMetadataFilesToKeep,
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java:553]: Why is this needed now?

File: server/src/main/java/org/opensearch/index/remote/RemoteStoreUtils.java (hunk 1 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/remote/RemoteStoreUtils.java]: this is already merged . I can take it up once remote translog and segment changes are merged .

File: server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java (hunk 3 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java]: we would need to delete all files except  the first `lastNMetadataFilesToKeep` .

File: server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java (hunk 2 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java]: +1 . this will part of remote segment PR now.

**Comment by ashking94** [comment]
`<inverted_primary_term>-- <inverted_generation>--<inverted_timestamp>--<primary_term>--`  - Why is primary_term needed here?

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java [hunk 4]
```diff
@@ -341,24 +359,34 @@ public void onFailure(Exception e) {
...
+                ThreadPool.Names.REMOTE_PURGE,
+                remoteMetadataTransferPath,
+                Integer.MAX_VALUE,
+                new ActionListener<>() {
+                    @Override
+                    public void onResponse(List<BlobMetadata> blobMetadata) {
+                        List<String> sortedMetadataFiles = blobMetadata.stream().map(BlobMetadata::name).collect(Collectors.toList());
+                        if (sortedMetadataFiles.size() <= 1) {
+                            logger.trace("Remote Metadata file count is {}, so skipping deletion", sortedMetadataFiles.size());
+                            return;
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:374]: before return we should be releasing the lock here.

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java [hunk 5]
```diff
@@ -341,24 +359,34 @@ public void onFailure(Exception e) {
...
+                    }
 
-            @Override
-            public void onFailure(Exception e) {
-                logger.error("Exception occurred while listing translog metadata files from remote store", e);
-            }
-        });
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.error("Exception occurred while listing translog metadata files from remote store", e);
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:383]: we should be releasing the permit here as well.

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java [hunk 6]
```diff
@@ -341,24 +359,34 @@ public void onFailure(Exception e) {
...
+                    }
+                }
+            );
+        } catch (Exception e) {
+            logger.error("Exception occurred while listing translog metadata files from remote store", e);
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:388]: we need to release the permit here also.

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java [hunk 7]
```diff
@@ -341,24 +359,34 @@ public void onFailure(Exception e) {
...
+        try {
+            transferService.listAllInSortedOrderAsync(
+                ThreadPool.Names.REMOTE_PURGE,
+                remoteMetadataTransferPath,
+                Integer.MAX_VALUE,
+                new ActionListener<>() {
+                    @Override
+                    public void onResponse(List<BlobMetadata> blobMetadata) {
+                        List<String> sortedMetadataFiles = blobMetadata.stream().map(BlobMetadata::name).collect(Collectors.toList());
+                        if (sortedMetadataFiles.size() <= 1) {
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:372]: this condition should not be true because there are generations to be deleted, and if that is true then this means that there will be atleast 1 metadata file to delete, right? Not expecting you change this, but lets think once if this is needed.

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java [hunk 8]
```diff
@@ -185,15 +185,39 @@ private void downloadToFS(String fileName, Path location, String primaryTerm) th
...
+                exceptionSetOnce.set((IOException) e);
+            }),
+            latch
+        );
+
+        try {
+            transferService.listAllInSortedOrder(remoteMetadataTransferPath, 1, latchedActionListener);
+            latch.await();
+        } catch (InterruptedException | IOException e) {
+            throw new IOException("Exception while reading/downloading metadafile", e);
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:213]: We are consuming most of the exceptions in `exceptionSetOnce` itself. When would we see exception thrown otherwise?

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferMetadata.java [hunk 9]
```diff
@@ -75,8 +76,20 @@ public Map<String, String> getGenerationToPrimaryTermMapper() {
...
+     */
+    public String getFileName() {
+        return String.join(
+            METADATA_SEPARATOR,
+            Arrays.asList(
+                RemoteStoreUtils.invertLong(primaryTerm),
+                RemoteStoreUtils.invertLong(generation),
+                RemoteStoreUtils.invertLong(createdAt),
+                String.valueOf(primaryTerm),
+                String.valueOf(generation)
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferMetadata.java:90]: Why do we need these when we have the inverted version present?

**Review by ashking94 — CHANGES_REQUESTED** [review]
Reviewed all except tests. will cover it in next review.

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java (hunk 7 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:372]: In case the tlog files get deleted , but md file deletion fails, this can still happen in future deletes.

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferMetadata.java (hunk 9 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferMetadata.java:90]: Makes sense. Removing it  in next commit.

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java (hunk 8 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:213]: we are throwing it later after the read/download finishes .

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java [hunk 10]
```diff
@@ -418,8 +419,6 @@ private void deleteStaleRemotePrimaryTermsAndMetadataFiles() {
             assert readers.isEmpty() == false : "Expected non-empty readers";
             long minimumReferencedPrimaryTerm = readers.stream().map(BaseTranslogReader::getPrimaryTerm).min(Long::compare).get();
             translogTransferManager.deletePrimaryTermsAsync(minimumReferencedPrimaryTerm);
-            // Second we delete all stale metadata files from remote store
-            translogTransferManager.deleteStaleTranslogMetadataFilesAsync();
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:422]: nit: If we are not deleting metadata files in this method, we can rename is accordingly.

**Comment by sachinpkale** [comment]
Changes LGTM. Re-triggering the build.

**Comment by gbbafna** [comment]
The failings tests are all flaky . 

```
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/11_indices_metrics/Metric - multi}
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=cat.nodes/10_basic/Test cat nodes output}
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/40_store_stats/Store stats}
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/11_indices_metrics/Metric - segments include_segment_file_sizes}
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/11_indices_metrics/Metric - recovery}
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/11_indices_metrics/Metric - _all}
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/20_response_filtering/Nodes Stats with response filtering}
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/30_discovery/Discovery stats}
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/11_indices_metrics/Metric - blank}
    org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/50_indexing_pressure/Indexing pressure stats}
    org.opensearch.remotestore.RemoteStoreRefreshListenerIT.testRemoteRefreshRetryOnFailure
    org.opensearch.remotestore.SegmentReplicationRemoteStoreIT.testCancellation
    org.opensearch.remotestore.SegmentReplicationRemoteStoreIT.testCancellation
    org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testNodeDropWithOngoingReplication
```

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java (hunk 10 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:422]: Done.

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java [hunk 11]
```diff
@@ -185,15 +185,39 @@ private void downloadToFS(String fileName, Path location, String primaryTerm) th
...
+            }, e -> {
+                logger.error(() -> new ParameterizedMessage("Exception while listing metadata files "), e);
+                exceptionSetOnce.set((IOException) e);
+            }),
+            latch
+        );
+
+        try {
+            transferService.listAllInSortedOrder(remoteMetadataTransferPath, 1, latchedActionListener);
+            latch.await();
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:211]: Do we need to keep a max timeout?

File: server/src/main/java/org/opensearch/index/translog/transfer/TransferService.java [hunk 12]
```diff
@@ -114,4 +107,8 @@ void uploadBlobAsync(
      */
     InputStream downloadBlob(Iterable<String> path, String fileName) throws IOException;
 
+    void listAllInSortedOrder(Iterable<String> path, int limit, ActionListener<List<BlobMetadata>> listener) throws IOException;
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/translog/transfer/TransferService.java:110]: Do we need to pass the `ActionListener` for non-async cases

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java (hunk 11 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:211]: I would rather rely on Blob Store to timeout using idle connection or do retries internally . Because if we timeout , shard is going to fail and is going to get retried .

File: server/src/main/java/org/opensearch/index/translog/transfer/TransferService.java (hunk 12 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/translog/transfer/TransferService.java:110]: Non-async will need to use `Latches` and make it async at the caller side.  This is consistent with `BlobContainer` interface as well .  If we are to provide a method of non async cases, BlobContainer would be the right place for this IMO .

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java (hunk 8 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:213]: Can we double check what all exceptions can be thrown here?

File: server/src/test/java/org/opensearch/index/translog/RemoteFSTranslogTests.java [hunk 13]
```diff
@@ -592,7 +592,7 @@ public void testMetadataFileDeletion() throws Exception {
         );
         translog.setMinSeqNoToKeep(totalDocs);
         translog.trimUnreferencedReaders();
-        assertBusy(() -> assertEquals(2, blobStoreTransferService.listAll(getTranslogDirectory().add(METADATA_DIR)).size()));
+        assertBusy(() -> assertEquals(1, blobStoreTransferService.listAll(getTranslogDirectory().add(METADATA_DIR)).size()));
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/index/translog/RemoteFSTranslogTests.java:595]: So, due to the new change, if the `translog.setMinSeqNoToKeep()` is set correctly, then we only hold maximum 1 metadata file?

File: server/src/test/java/org/opensearch/index/translog/RemoteFSTranslogTests.java [hunk 14]
```diff
@@ -607,10 +607,6 @@ public void testMetadataFileDeletion() throws Exception {
         long oldPrimaryTerm = primaryTerm.get();
         long newPrimaryTerm = primaryTerm.incrementAndGet();
 
-        // Check all metadata files corresponds to old primary term
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/index/translog/RemoteFSTranslogTests.java:610]: Why have we removed this check on primary term names at 2 places in this file?

File: server/src/test/java/org/opensearch/index/translog/transfer/TranslogTransferManagerTests.java [hunk 15]
```diff
@@ -186,66 +191,92 @@ public void testReadMetadataNoFile() throws IOException {
...
+            latchedActionListener.onResponse(bmList);
+            return null;
+        }).when(transferService).listAllInSortedOrder(any(BlobPath.class), anyInt(), any(ActionListener.class));
 
-        when(transferService.listAll(remoteBaseTransferPath)).thenReturn(Sets.newHashSet());
         assertNull(translogTransferManager.readMetadata());
     }
 
-    public void testReadMetadataSingleFile() throws IOException {
+    public void testReadMetadataHappy() throws IOException {
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/index/translog/transfer/TranslogTransferManagerTests.java:204]: nit - let's rename it to `testReadMetadata()`? (and a java doc saying it's for happy case :) )

**Review by ashking94 — CHANGES_REQUESTED** [review]
lets add UTs/ITs for metadata file deletion?

File: server/src/test/java/org/opensearch/index/translog/RemoteFSTranslogTests.java (hunk 14 above)
Comment (gbbafna) [reply server/src/test/java/org/opensearch/index/translog/RemoteFSTranslogTests.java:610]: It doesn't contain primary term name now .  Hence , it is no longer needed .

File: server/src/test/java/org/opensearch/index/translog/RemoteFSTranslogTests.java (hunk 13 above)
Comment (gbbafna) [reply server/src/test/java/org/opensearch/index/translog/RemoteFSTranslogTests.java:595]: Yes , that is correct.

**Review by sachinpkale — APPROVED** [review]

**Review by ashking94 — APPROVED** [review]
